contracts/smart-contract-wallet/modules/SessionKeyModule.sol

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {BaseAuthorizationModule, UserOperation, ISignatureValidator} from "./BaseAuthorizationModule.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@account-abstraction/contracts/core/Helpers.sol";
+import "./SessionValidationModules/ISessionValidationModule.sol";
+
+struct SessionStorage {
+    bytes32 merkleRoot;
+}
+
+contract SessionKeyManager is BaseAuthorizationModule {
+    /*
+     * @dev mapping of Smart Account to a SessionStorage
+     * Info about session keys is stored as root of the merkle tree built over the session keys
+     */
+    mapping(address => SessionStorage) internal userSessions;
+
+    function getSessionKeys(
+        address smartAccount
+    ) external view returns (SessionStorage memory) {
+        return userSessions[smartAccount];
+    }
+
+    function setMerkleRoot(bytes32 _merkleRoot) external {
+        _setSessionData(msg.sender, _merkleRoot);
+    }
+
+    /**
+     * @dev validates userOperation
+     * @param userOp User Operation to be validated.
+     * @param userOpHash Hash of the User Operation to be validated.
+     * @return sigValidationResult 0 if signature is valid, SIG_VALIDATION_FAILED otherwise.
+     */
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) external view virtual returns (uint256) {
+        SessionStorage storage sessionKeyStorage = _getSessionData(msg.sender);
+        (bytes memory moduleSignature, ) = abi.decode(
+            userOp.signature,
+            (bytes, address)
+        );
+        (
+            uint48 validUntil,
+            uint48 validAfter,
+            address sessionValidationModule,
+            bytes memory sessionKeyData,
+            bytes32[] memory merkleProof,
+            bytes memory sessionKeySignature
+        ) = abi.decode(
+                moduleSignature,
+                (uint48, uint48, address, bytes, bytes32[], bytes)
+            );
+
+        bytes32 leaf = keccak256(
+            abi.encodePacked(
+                validUntil,
+                validAfter,
+                sessionValidationModule,
+                sessionKeyData
+            )
+        );
+        if (
+            !MerkleProof.verify(merkleProof, sessionKeyStorage.merkleRoot, leaf)
+        ) {
+            revert("SessionNotApproved");
+        }
+        return
+            _packValidationData(
+                //_packValidationData expects true if sig validation has failed, false otherwise
+                !ISessionValidationModule(sessionValidationModule)
```

Comment:
do we need to handle reverts here by making low level call and then control flow?

Reply:
hmm we could return SIG_VALIDATION_FAILED from validation module or catch the error in validateUserOp and return 1. because currently it would just revert even with say wrong passkey/ecdsa sig and won't allow backend systems  as bundler node to call simulateHandleOp with dummysig etc

Reply:
The solution here would be to rename SIG_VALIDATION_FAILED to just VALIDATION_FAILED

Reply:
@arcticfloyd1984 ^^^

Reply:
>currently it would just revert even with say wrong passkey/ecdsa sig

nope. In case of a wrong ecdsa sig it won't revert. in this case it will return SIG_VALIDATION_FAILED

https://github.com/bcnmy/scw-contracts/blob/a6ed50f2fdb317a6fcf5eff2e85749838fe12ce2/contracts/smart-contract-wallet/modules/SessionValidationModules/ERC20SessionValidationModule.sol#L58-L63

Reply:
> The solution here would be to rename SIG_VALIDATION_FAILED to just VALIDATION_FAILED Then it will be good to return it in any case when validation fails
> 
> however, it makes things less informative. as it provides no information about the failure reason. that's why I decided to revert in ERC20SessionValidationModule and also revert with "SessionNotApproved" in SessionKeyManager

hmm I think then in this case we can leave it with reverts which will be captured as AA23. I just realised in paymasters repo and flow also there is AA33 expected for all cases not related to sig verification.

Reply:
> won't allow backend systems as bundler node to call simulateHandleOp with dummysig etc

however, EP.handleOps will revert in both cases, see [here](https://github.com/bcnmy/scw-contracts/pull/98#discussion_r1239682277)

so again, it makes no difference in terms of simulating handleOps whether to return sig_validation_failed or revert, they both lead EP to revert